Flow.scanAsync does not complete after the initial value has been pulled.

# The Problem
The `Flow.scanAsync` stage does not propagate completion to downstream stages after the initial value (zero) has been pulled.

# Test
```scala
implicit val system: ActorSystem =
  ActorSystem()

implicit val mat: Materializer =
  ActorMaterializer()

val flow: Flow[Int, Int, NotUsed] =
 Flow[Int].scanAsync(0)((acc, in) => Future.successful(acc + in))

val (pub, sub) = TestSource
 .probe[Int]
 .via(flow)
 .toMat(TestSink.probe)(Keep.both)
 .run()

sub.request(10) // signal demand
sub.expectNext(0) // expect initial value
pub.sendComplete() // signal completion
sub.expectComplete() // fails due to time out!
```
This test fails with:
`java.lang.AssertionError: assertion failed: timeout (6 seconds) during expectMsg while waiting for OnComplete`

# Possible cause
The class `akka.stream.impl.fusing.ScanAsync` overrides the default implementation of the `onUpstreamFinish` method with an [empty implementation](https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala#L510).

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowScanAsyncSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowScanAsyncSpec.scala
@@ -3,10 +3,11 @@
  */
 package akka.stream.scaladsl
 
-import akka.pattern
+import akka.{ NotUsed, pattern }
 import akka.stream.{ ActorAttributes, ActorMaterializer, Supervision }
 import akka.stream.impl.ReactiveStreamsCompliance
 import akka.stream.testkit.TestSubscriber.Probe
+import akka.stream.testkit.Utils.TE
 import akka.stream.testkit._
 import akka.stream.testkit.scaladsl._
 
@@ -31,6 +32,32 @@ class FlowScanAsyncSpec extends StreamSpec {
         .runWith(TestSink.probe[Int])
         .request(1)
         .expectNextOrComplete(0)
+    }
+
+    "complete after zero-element has been consumed" in {
+      val (pub, sub) =
+        TestSource.probe[Int]
+          .via(Flow[Int].scanAsync(0)((acc, in) ⇒ Future.successful(acc + in)))
+          .toMat(TestSink.probe)(Keep.both)
+          .run()
+
+      sub.request(10)
+      sub.expectNext(0)
+      pub.sendComplete()
+      sub.expectComplete()
+    }
+
+    "fail after zero-element has been consumed" in {
+      val (pub, sub) =
+        TestSource.probe[Int]
+          .via(Flow[Int].scanAsync(0)((acc, in) ⇒ Future.successful(acc + in)))
+          .toMat(TestSink.probe)(Keep.both)
+          .run()
+
+      sub.request(10)
+      sub.expectNext(0)
+      pub.sendError(TE("bang"))
+      sub.expectError(TE("bang"))
     }
 
     "work with a single source" in {

--- a/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
+++ b/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
@@ -432,7 +432,8 @@ private[stream] object Collect {
       private lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
 
       private val ZeroHandler: OutHandler with InHandler = new OutHandler with InHandler {
-        override def onPush(): Unit = ()
+        override def onPush(): Unit =
+          throw new IllegalStateException("No push should happen before zero value has been consumed")
 
         override def onPull(): Unit = {
           push(out, current)
@@ -507,7 +508,16 @@ private[stream] object Collect {
         }
       }
 
-      override def onUpstreamFinish(): Unit = {}
+      override def onUpstreamFinish(): Unit = {
+        if (current == zero) {
+          eventualCurrent.value match {
+            case Some(Success(`zero`)) ⇒
+              // #24036 upstream completed without emitting anything but after zero was emitted downstream
+              completeStage()
+            case _ ⇒ // in all other cases we will get a complete when the future completes
+          }
+        }
+      }
 
       override val toString: String = s"ScanAsync.Logic(completed=${eventualCurrent.isCompleted})"
     }

